walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,19 +15,137 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Comparator;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
+
+  /*
+   * Given a list of scheduled {@code events}, query() returns a list of all available times for the 
+   * new meeting {@code request}.
+   *
+   * @param events: list of scheduled events for attendees/optional attendees
+   * @param request: the new meeting that is to be scheduled 
+   *
+   * @return: a list of all available times the meeting can be held
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-    //create an empty Collection<TimeRange> for available times for the requested meeting
-    //check if the duration of the event is more than a day or if the collection of attendees is empty
-    //if yes, return the empty Collection
-    //if no, begin looking for appropriate time slots
-    //Iterate through the day by checking each appropriate time duration ex. if meeting duration is 1hr 8-9, 8:30-9:30, ...
-    //for each time duration check if there is an event present and if anyone is already attending that meeting
-    //If the time is accepted, add it to the collection
-    //return the collection of times
+    
+    /* Checks if the new meeting {@code request} is longer than a day.
+     * If yes, return an empty list;
+     */
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return new ArrayList<>();
+    }
+
+    /* Checks if there are optional attendees for the {@code request}
+     * If yes, search for a time where both optional and non-optional attendees can attend.
+     * If no such time exists, priority goes to non-optional attendees
+     */
+    if(request.checkForOptionalAttendees()){
+      ArrayList<TimeRange> availableTimes = findAvailableMeetingTimes(events, request, true);
+      if(!request.checkForAttendees() || !availableTimes.isEmpty()){
```

Comment:
Yes that is also my understanding of the problem.

Reply:
I removed the first part of the conditional and all tests seem to pass. I left that part in because I thought that for the case where there are optional attendees with no available time slots they can all attend and non-optional attendees that need to attend, the !request.checkForAttendees() covered this case. But I see that my check for whether the availableTimes.isEmpty() covers that case.